### PR TITLE
feat: LSP inlay hints for let and global

### LINKED ISSUE #5513: LSP: support inlay type hints for `let` and `global`
### Problem

It would be nice if inlay type hints were shown for `let` and `global` similar to how it works in Rust Analyzer. For example, this:

```rust
let var = 0;
```

would add an inlay type hint so the code looks like this:

```rust
let var: Field = 0;
```

### Happy Case

The editor shows inlay type hints in the visible code region.

### Project Impact

None

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #5513

## Summary

Shows type hints for `let` and `global`. Note that right now they will always show for `let` and `global` that don't have an explicit type. Eventually it would be nice to have a configuration to toggle them, but the noise they introduce isn't that much so we could do that in a separate PR (also paving the way for configuring other types of inlay hints, like parameter names or names next to closing braces).

https://github.com/user-attachments/assets/4b743873-e1bf-430b-8632-4020ea53fbd8

## Additional Context

If you try it, hovering over types might not work well because of #5515, so you could merge that branch into this one before testing things.

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/lsp/src/requests/inlay_hint.rs [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+            let mut collector = InlayHintCollector::new(args, file_id, span);
+            collector.collect_in_parsed_module(&parsed_moduled);
+            collector.inlay_hints
+        })
+    });
+    future::ready(result)
+}
+
+pub(crate) struct InlayHintCollector<'a> {
+    args: ProcessRequestCallbackArgs<'a>,
```
Comment (TomAFrench) [inline tooling/lsp/src/requests/inlay_hint.rs:69]: We only use `interner` and `files` from this right? Can we pull out these fields from `args` so it's clearer on what we need?

**Review by TomAFrench — COMMENTED** [review]
Looks good, this is one of the LSP features I've been wanting the most so very happy to see it implemented!

Some small nits and then we can merge.

File: tooling/lsp/src/requests/inlay_hint.rs [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+
+        let mut parts = Vec::new();
+        parts.push(str_part(": "));
+        self.push_type_parts(typ, &mut parts);
+
+        self.inlay_hints.push(InlayHint {
+            position,
+            label: InlayHintLabel::LabelParts(parts),
+            kind: Some(InlayHintKind::TYPE),
+            text_edits: None,
```
[the remark this answers is not in the document]
Comment (asterite) [reply tooling/lsp/src/requests/inlay_hint.rs:328]: TIL you can double-click on them to insert them!

I'll capture an issue. I just tried implementing it and it works, but when it's inserted the old inlay hint is still there... I checked Rust Analyzer and when that happens it also sends a "refresh inlay hints" request, and doing that is a bit more work so better as a separate PR.

Captured in #5527

File: tooling/lsp/src/requests/inlay_hint.rs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+    }
+}
+
+fn string_part(str: String) -> InlayHintLabelPart {
+    InlayHintLabelPart { value: str, location: None, tooltip: None, command: None }
+}
+
+fn str_part(str: &str) -> InlayHintLabelPart {
+    string_part(str.to_string())
+}
```
[the remark this answers is not in the document]
Comment (asterite) [reply tooling/lsp/src/requests/inlay_hint.rs:483]: Thank you! I'm still new to Rust so these suggestions are really helpful.

**Review by TomAFrench — APPROVED** [review]
LGTM!

**Pull request merged**